kasa/deviceconfig.py

```diff
@@ -214,22 +175,30 @@ def __post_init__(self) -> None:
                 DeviceFamily.IotSmartPlugSwitch, DeviceEncryptionType.Xor
             )
 
-    def to_dict(
+    def __pre_serialize__(self) -> Self:
+        return replace(self, http_client=None)
+
+    def to_dict_control_credentials(
         self,
         *,
         credentials_hash: str | None = None,
         exclude_credentials: bool = False,
     ) -> dict[str, dict[str, str]]:
-        """Convert device config to dict."""
-        if credentials_hash is not None or exclude_credentials:
-            self.credentials = None
-        if credentials_hash:
-            self.credentials_hash = credentials_hash
-        return _dataclass_to_dict(self)
+        """Convert deviceconfig to dict controlling how to serialize credentials.
```

Comment:
Do we have a real use case for not using the default and avoid having extra code to strip this?

Reply:
We don't really have a case ourselves anymore as we don't serialize the whole `deviceconfig` in HA anymore. Maybe we could remove this in a subsequent PR or mark it for deprecation. Also perhaps the default should be to not serialize the credentials as well?

Reply:
Yeah, I think we could remove it in the future, but it's not that critical. I would perhaps avoid making it too magical, but I'd guess skipping pw&user in favor of credentials_hash makes sense to avoid storing clear text creds (but would break on protocol updates).